cpp/src/lib/manifest.cc

```diff
@@ -205,22 +202,127 @@ template <> void from2(message &m, Manifest &mf, bool &ok_to_continue) {
+    fragment_checksums[c] = std::move(chunk);
   }
 
+  layout<uint32_t> fragment_packed_sizes;
   uint8_t layout_tag3;
   from(m2, layout_tag3);
   if (layout_tag3 != 1) {
     throw deserialisation_exception("unexpected layout tag 3");
   }
 
-  from(m2, mf.fragment_packed_sizes);
+  from(m2, fragment_packed_sizes);
+
+  // build mf.fragments
+  for (uint32_t c = 0; c < n_chunks; c++) {
+    uint32_t n_fragments = fragment_locations[c].size();
+    std::vector<std::shared_ptr<Fragment>> chunk;
+    for (uint32_t f = 0; f < n_fragments; f++) {
+      std::shared_ptr<Fragment> fragment_ptr(new Fragment());
+      fragment_ptr->loc = fragment_locations[c][f];
+      fragment_ptr->crc = fragment_checksums[c][f];
+      fragment_ptr->len = fragment_packed_sizes[c][f];
+      chunk.push_back(fragment_ptr);
+    }
+    mf.fragments.push_back(std::move(chunk));
+  }
 
   from(m2, mf.version_id);
   from(m2, mf.max_disks_per_node);
   from(m2, mf.timestamp);
 }
 
+template <> void from(message &m, Fragment &f) {
+  uint32_t fragment_s_size;
+  from(m, fragment_s_size);
+  auto m2 = m.get_nested_message(fragment_s_size);
+  m.skip(fragment_s_size);
+
+  uint8_t version;
+  from(m2, version);
+  if (version != 1) {
+    throw deserialisation_exception("unexpected Fragment version");
+  }
+  from(m2, f.loc);
+
+  Checksum *p;
+  from(m2, p);
+  std::shared_ptr<Checksum> sp(p);
+  f.crc = sp;
+
+  from(m2, f.len);
+}
+
+void _from_version2(message &m, Manifest &mf, bool &ok_to_continue) {
+  ALBA_LOG(DEBUG, "_from_version2");
+  uint32_t compressed_size;
+  from(m, compressed_size);
+  auto m_compressed = m.get_nested_message(compressed_size);
+  m.skip(compressed_size);
+
+  std::string real;
+  snappy::Uncompress(m_compressed.current(compressed_size), compressed_size,
+                     &real);
+
+  auto buffer = message_buffer::from_string(real); // TODO:copies
+  ok_to_continue = true;
+  message m2(buffer);
+  from(m2, mf.name);
+  from(m2, mf.object_id);
+  from(m2, mf.chunk_sizes);
+
+  uint8_t version2;
+  from(m2, version2);
+  if (version2 != 1) {
+    throw deserialisation_exception("unexpected version2");
+  }
+
+  from(m2, mf.encoding_scheme);
+  from(m2, mf.compression);
+  from(m2, mf.encrypt_info);
+  from(m2, mf.checksum);
+  from(m2, mf.size);
+  uint8_t layout_tag;
+  from(m2, layout_tag);
+  if (layout_tag != 1) {
+    throw deserialisation_exception("unexpected layout tag");
+  }
+  // from(m2, mf.fragments);
+
+  uint32_t n_chunks;
+  from(m2, n_chunks);
+  mf.fragments.resize(n_chunks);
+
+  for (int32_t c = n_chunks - 1; c >= 0; --c) {
+    uint32_t n_fragments;
+    from(m2, n_fragments);
+    std::vector<std::shared_ptr<Fragment>> chunk(n_fragments);
+    for (int32_t f = n_fragments - 1; f >= 0; --f) {
+      std::shared_ptr<Fragment> fragment_ptr(new Fragment());
+      from(m2, *fragment_ptr);
+      chunk[f] = std::move(fragment_ptr);
+    };
+    mf.fragments[c] = std::move(chunk);
```

Comment:
I don't think (obviously backed by the test) it's the wrong order.
https://github.com/openvstorage/alba/blob/master/cpp/include/llio.h#L262